src/stan/io/writer.hpp

```diff
@@ -279,81 +282,148 @@ class writer {
    */
   void matrix_unconstrain(const matrix_t &y) {
     typedef typename stan::math::index_type<matrix_t>::type idx_t;
-    for (idx_t j = 0; j < y.cols(); ++j)
-      for (idx_t i = 0; i < y.rows(); ++i)
+    for (idx_t j = 0; j < y.cols(); ++j) {
+      for (idx_t i = 0; i < y.rows(); ++i) {
         data_r_.push_back(y(i, j));
+      }
+    }
+  }
+
+  /**
+   * Write the specified unconstrained sparse matrix.
+   *
+   * @param y Sparse Matrix to write.
+   */
+  void sparse_matrix_unconstrain(sparse_matrix_t &y) {
+    for (int k = 0; k < y.outerSize(); ++k) {
+      for (sparse_matrix_inner_iter_t it(y, k); it; ++it) {
+        data_r_.push_back(it.value());
+      }
+    }
   }
 
   void vector_lb_unconstrain(double lb, vector_t &y) {
     typedef typename stan::math::index_type<vector_t>::type idx_t;
-    for (idx_t i = 0; i < y.size(); ++i)
+    for (idx_t i = 0; i < y.size(); ++i) {
       scalar_lb_unconstrain(lb, y(i));
+    }
   }
   void row_vector_lb_unconstrain(double lb, row_vector_t &y) {
     typedef typename stan::math::index_type<row_vector_t>::type idx_t;
-    for (idx_t i = 0; i < y.size(); ++i)
+    for (idx_t i = 0; i < y.size(); ++i) {
       scalar_lb_unconstrain(lb, y(i));
+    }
   }
   void matrix_lb_unconstrain(double lb, matrix_t &y) {
     typedef typename stan::math::index_type<matrix_t>::type idx_t;
-    for (idx_t j = 0; j < y.cols(); ++j)
-      for (idx_t i = 0; i < y.rows(); ++i)
+    for (idx_t j = 0; j < y.cols(); ++j) {
+      for (idx_t i = 0; i < y.rows(); ++i) {
         scalar_lb_unconstrain(lb, y(i, j));
+      }
+    }
+  }
+
+  void sparse_matrix_lb_unconstrain(double lb, sparse_matrix_t &y) {
+    for (int k = 0; k < y.outerSize(); ++k) {
+      for (sparse_matrix_inner_iter_t it(y, k); it; ++it) {
+        scalar_lb_unconstrain(lb, it.value());
+      }
+    }
   }
 
   void vector_ub_unconstrain(double ub, vector_t &y) {
     typedef typename stan::math::index_type<vector_t>::type idx_t;
-    for (idx_t i = 0; i < y.size(); ++i)
+    for (idx_t i = 0; i < y.size(); ++i) {
       scalar_ub_unconstrain(ub, y(i));
+    }
   }
   void row_vector_ub_unconstrain(double ub, row_vector_t &y) {
     typedef typename stan::math::index_type<row_vector_t>::type idx_t;
-    for (idx_t i = 0; i < y.size(); ++i)
+    for (idx_t i = 0; i < y.size(); ++i) {
       scalar_ub_unconstrain(ub, y(i));
+    }
   }
   void matrix_ub_unconstrain(double ub, matrix_t &y) {
     typedef typename stan::math::index_type<matrix_t>::type idx_t;
-    for (idx_t j = 0; j < y.cols(); ++j)
-      for (idx_t i = 0; i < y.rows(); ++i)
+    for (idx_t j = 0; j < y.cols(); ++j) {
+      for (idx_t i = 0; i < y.rows(); ++i) {
         scalar_ub_unconstrain(ub, y(i, j));
+      }
+    }
+  }
+
+  void sparse_matrix_ub_unconstrain(double ub, sparse_matrix_t &y) {
```

Comment:
Documentation.

Also, is there no way to get a single iterator in the right order withouthaving to double stride like this through the ouer and inner bits?

Again, I want to see all this code cut-and-paste encapsulated in something more general if possible.